pkg/jobs/job_info_storage.go

```diff
@@ -8,18 +8,317 @@ package jobs
+	if !ok {
+		return 0, hlc.Timestamp{}, time.Time{}, errors.AssertionFailedf("job progress: expected value to be DTimestampTZ (was %T)", row[1])
+	}
+
+	var fraction float64
+	if row[1] == tree.DNull {
+		fraction = math.NaN()
+	} else {
+		fracDatum, ok := row[1].(*tree.DFloat)
+		if !ok {
+			return 0, hlc.Timestamp{}, time.Time{}, errors.AssertionFailedf("job progress: expected value to be DFloat (was %T)", row[1])
+		}
+		fraction = float64(*fracDatum)
+	}
+
+	var ts hlc.Timestamp
+	if row[2] != tree.DNull {
+		resolved, ok := row[2].(*tree.DDecimal)
+		if !ok {
+			return 0, hlc.Timestamp{}, time.Time{}, errors.AssertionFailedf("job progress: expected value to be DDecimal (was %T)", row[1])
+		}
+		ts, err = hlc.DecimalToHLC(&resolved.Decimal)
+		if err != nil {
+			return 0, hlc.Timestamp{}, time.Time{}, err
+		}
+	}
+
+	return fraction, ts, written.Time, nil
+}
+
+// Set records a progress update. If fraction is NaN or resolved is empty, that
+// field is left null. The time at which the progress was reported is recorded.
+func (i ProgressStorage) Set(
+	ctx context.Context, txn isql.Txn, fraction float64, resolved hlc.Timestamp,
+) error {
+	ctx, sp := tracing.ChildSpan(ctx, "write-job-progress")
+	defer sp.Finish()
+
+	if _, err := txn.ExecEx(
+		ctx, "write-job-progress-delete", txn.KV(), sessiondata.NodeUserSessionDataOverride,
+		`DELETE FROM system.job_progress WHERE job_id = $1`, i,
+	); err != nil {
+		return err
+	}
+
+	var frac, ts interface{}
+	if !math.IsNaN(fraction) {
+		frac = fraction
+	}
+	if resolved.IsSet() {
+		ts = resolved.AsOfSystemTime()
+	}
+
+	if _, err := txn.ExecEx(
+		ctx, "write-job-progress-insert", txn.KV(), sessiondata.NodeUserSessionDataOverride,
+		`INSERT INTO system.job_progress (job_id, written, fraction, resolved) VALUES ($1, now(), $2, $3)`,
+		i, frac, ts,
+	); err != nil {
+		return err
+	}
+
+	if _, err := txn.ExecEx(
+		ctx, "write-job-progress-history-insert", txn.KV(), sessiondata.NodeUserSessionDataOverride,
+		`UPSERT INTO system.job_progress_history (job_id, written, fraction, resolved) VALUES ($1, now(), $2, $3)`,
+		i, frac, ts,
+	); err != nil {
+		return err
+	}
+
+	if _, err := txn.ExecEx(
+		ctx, "write-job-progress-history-prune", txn.KV(), sessiondata.NodeUserSessionDataOverride,
+		`DELETE FROM system.job_progress_history WHERE job_id = $1 AND written IN 
+			(SELECT written FROM system.job_progress_history WHERE job_id = $1 ORDER BY written DESC OFFSET $2)`,
+		i, retainedProgressHistory.Get(txn.KV().DB().SettingsValues()),
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// StatusStorage reads and writes status rows and messages, i.e. short strings
+// that are for *human* consumption. These are distinct from info rows, which
+// are byte strings used by the code of the job itself to persist state that it
+// will consume, and distinct from the message log, as a job can only have one
+// "status" at a time, but messages are histoical log.
```

Comment:
it may be worth clarifying here that StatusStorage will only store running status messages, and not the job state (e.g. running vs paused).